Dockerfile

```
@@ -13,105 +13,148 @@
-RUN go build -o torchserve-adapter model-mesh-torchserve-adapter/main.go
+# https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope
+# don't provide "default" values (e.g. 'ARG TARGETARCH=amd64') for non-buildx environments,
+# see https://github.com/docker/buildx/issues/510
+ARG TARGETOS
+ARG TARGETARCH
+
+ARG GOOS=${TARGETOS:-linux}
+ARG GOARCH=${TARGETARCH:-amd64}
+
+# Build the binaries using native go compiler from BUILDPLATFORM but compiled output for TARGETPLATFORM
+# https://www.docker.com/blog/faster-multi-platform-builds-dockerfile-cross-compilation-guide/
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/go/pkg \
+    go build -o puller model-serving-puller/main.go && \
+    go build -o puller model-serving-puller/main.go && \
+    go build -o triton-adapter model-mesh-triton-adapter/main.go && \
+    go build -o mlserver-adapter model-mesh-mlserver-adapter/main.go && \
+    go build -o ovms-adapter model-mesh-ovms-adapter/main.go && \
+    go build -o torchserve-adapter model-mesh-torchserve-adapter/main.go
+
 
 ###############################################################################
 # Stage 3: Copy build assets to create the smallest final runtime image
 ###############################################################################
-FROM  registry.access.redhat.com/ubi8/ubi-minimal:8.4 as runtime
+FROM registry.access.redhat.com/ubi8/ubi-minimal:8.7 as runtime
 
-ARG IMAGE_VERSION
-ARG COMMIT_SHA
 ARG USER=2000
 
-LABEL name="model-serving-runtime-adapter" \
-      version="${IMAGE_VERSION}" \
-      release="${COMMIT_SHA}" \
-      summary="Sidecar container which runs in the Model-Mesh Serving model server pods" \
-      description="Container which runs in each model serving pod and act as an intermediary between model-mesh and third-party model-server containers"
-
 USER root
+
 # install python to convert keras to tf
-RUN microdnf install \
-    gcc \
-    gcc-c++ \
-    python38 && \
-    ln -sf /usr/bin/python3 /usr/bin/python && \
-    ln -sf /usr/bin/pip3 /usr/bin/pip && \
-    pip install tensorflow
+# NOTE: tensorflow not supported on PowerPC (ppc64le) or System Z (s390x) https://github.com/tensorflow/tensorflow/issues/46181
+RUN --mount=type=cache,target=/root/.cache/microdnf:rw \
+    microdnf install --setopt=cachedir=/root/.cache/microdnf \
+       gcc \
+       gcc-c++ \
+       python38-devel \
+       python38 \
+    && ln -sf /usr/bin/python3 /usr/bin/python \
+    && ln -sf /usr/bin/pip3 /usr/bin/pip \
+    && true
+
+# need to upgrade pip and install wheel before installing grpcio, before installing tensorflow on aarch64
+# use caching to speed up multi-platform builds
+ARG PIP_CACHE_DIR=/root/.cache/pip
+RUN --mount=type=cache,target=$PIP_CACHE_DIR \
+    pip install --upgrade pip && \
+    pip install wheel && \
+    pip install grpcio && \
+    pip install tensorflow && \
+    pip list && \
+    pip cache info
```

NIT: assuming that these were only needed for debugging cacheing
```suggestion
    pip install tensorflow
```

---

Right, but since they don't take up disk space or cost much time overhead I left them in.

---

Removed it.